src/rust/lib/graph-editor/src/component/visualization/container.rs

```diff
@@ -0,0 +1,174 @@
+//! This module defines the `Container` struct and related functionality.
+
+use crate::prelude::*;
+
+use crate::frp;
+use crate::visualization::*;
+
+use ensogl::display::traits::*;
+use ensogl::display;
+
+
+
+// ===========
+// === FRP ===
+// ===========
+
+/// Event system of the `Container`.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub struct ContainerFrp {
+    pub network           : frp::Network,
+    pub set_visibility    : frp::Source<bool>,
+    pub toggle_visibility : frp::Source,
+    pub set_visualization : frp::Source<Option<Visualization>>,
+    pub set_data          : frp::Source<Option<Data>>,
+}
+
+impl Default for ContainerFrp {
+    fn default() -> Self {
+        frp::new_network! { visualization_events
+            def set_visibility    = source::<bool>                  ();
+            def toggle_visibility = source::<()>                    ();
+            def set_visualization = source::<Option<Visualization>> ();
+            def set_data          = source::<Option<Data>>          ();
```

Comment:
You are repeating types here. They are not needed. They are provided literally 5 lines above